### PR TITLE
feature(Training): bring over the essence of `drift`

### PR DESCRIPTION
**Pull request by itchingpixels**

### CODE REVIEW

File: all_types.py [hunk 1]
```diff
@@ -1,9 +1,9 @@
...
 ModelOverTime = pd.Series
-InSamplePredictions = Union[np.ndarray, pd.Series]
-OutSamplePredictions = Union[np.ndarray, pd.Series]
-y = Union[np.ndarray, pd.Series]
-X = Union[np.ndarray, pd.DataFrame]
+TransformationsOverTime = List[pd.Series]
+InSamplePredictions = pd.Series
+OutSamplePredictions = pd.Series
+y = pd.Series
+X = pd.DataFrame
```
Comment (itchingpixels) [inline all_types.py:9]: there's actually a difference here, X and y are kept as original, but a model always gets and returns an np.ndarray (see updated Model interface).
that's what all of the sklearn models do, which will become the most common denominator therefore.

File: models/sktime_wrapper.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
+from .base import Model
+import numpy as np
+from sktime.forecasting.base import ForecastingHorizon, BaseForecaster
+from all_types import OutSamplePredictions
+
+
+class SKTimeModelWrapper(Model):
```
Comment (itchingpixels) [inline models/sktime_wrapper.py:7]: generic sktime model wrapper

File: all_types.py (hunk 1 above)
Comment (szemyd) [reply all_types.py:9]: Should we keep the aliases? I thought we discussed that it might be harder to parse for others?

Comment (szemyd) [reply all_types.py:9]: but more concise and easier to read for us.

File: train/fit_transformations.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
+import pandas as pd
+from all_types import TransformationsOverTime, X, y
+from tqdm import tqdm
+from transformations.base import Transformation
+from utils.splitters import Split, Splitter
+from typing import List
+
+
+def fit_transformations(
```
Comment (itchingpixels) [inline train/fit_transformations.py:9]: I think this is the only thing that's a bit tricky. I can explain it

File: train/inference.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+            split,
+            X,
+            model_over_time,
+            transformations_over_time,
+        )
+        for split in tqdm(splitter.splits())
+    ]
+    for batch in batched_results:
+        for index, prediction, probs in batch:
+            insample_predictions.iloc[index] = prediction
```
Comment (itchingpixels) [inline train/inference.py:39]: I think this can be done in a more efficient manner

File: train/train.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    models = [
+        __train_on_window(split, X, y, model, transformations_over_time)
+        for split in tqdm(splitter.splits())
+    ]
+
+    models_over_time = pd.Series(index=y.index, dtype="object").rename(model.name)
+
+    # aggregate models
+    for index, current_model in models:
+        models_over_time.iloc[index] = current_model
```
Comment (itchingpixels) [inline train/train.py:27]: probably can be done more efficently

File: train/inference.py (hunk 4 above)
Comment (itchingpixels) [reply train/inference.py:39]: and probably I can clean this up a bit.

File: train/fit_transformations.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (itchingpixels) [reply train/fit_transformations.py:9]: no, we're fitting the transformations! then we do inference on the transformations on the fly, before training and model inference.
that's the important part, that we should discuss. there's another way, but it's way more memory-heavy

File: all_types.py (hunk 1 above)
Comment (itchingpixels) [reply all_types.py:9]: I think X and y are not important, but the rest are quite useful in identifying what a function gets as a parameter/returns.

File: transformations/fracdiff.py [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+    def __init__(self, window_size: int):
+        self.window_size = window_size
+
+    def fit(self, X: pd.DataFrame, y: Optional[pd.Series] = None) -> None:
+        self.model = FracdiffStat(window=self.window_size)
+        self.model.fit(X, y)
+
+    def transform(self, X: pd.DataFrame) -> pd.DataFrame:
+        X = pd.DataFrame(self.model.transform(X), index=X.index)
+        X.columns = ["Fracdiff_" + str(i) for i in range(1, len(X.columns) + 1)]
```
Comment (itchingpixels) [inline transformations/fracdiff.py:22]: this is wrong

File: all_types.py (hunk 1 above)
Comment (itchingpixels) [reply all_types.py]: removed X and y

File: models/sktime_naive_models.py [hunk 7]
```diff
@@ -54,5 +53,5 @@ def predict(self, X: X) -> OutSamplePredictions:
 
 
 default_naive_model = NaiveForecasterWrapper(
-    NaiveForecasterConfig(strategy=NaiveForecasterWrapper.strategy_types.last)
```
[the remark this answers is not in the document]
Comment (itchingpixels) [reply models/sktime_naive_models.py:57]: I think it's fine! I kept it in the end!

File: transformations/fracdiff.py (hunk 6 above)
Comment (itchingpixels) [reply transformations/fracdiff.py]: fixed

File: tests/test_inference.py [hunk 8]
```diff
@@ -0,0 +1 @@
+#(outofsample_predictions == outofsample_predictions.rename(insample_predictions.name)).sum()
```
Comment (itchingpixels) [inline tests/test_inference.py:1]: I'll convert this soon into a unit test

**Pull request merged**